cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/PreferredLeaderElectionGoal.java

```diff
@@ -38,10 +56,20 @@ public boolean optimize(ClusterModel clusterModel, Set<Goal> optimizedGoals, Set
     Set<TopicPartition> partitionsToMove = new HashSet<>();
     for (Broker b : clusterModel.demotedBrokers()) {
       for (Replica r : b.replicas()) {
-        Partition p = clusterModel.partition(r.topicPartition());
-        p.moveReplicaToEnd(r);
+        // There are two scenarios where replica swap operation is skipped:
+        // 1.the replica on the to-be-demoted broker is not leader replica and skip_reorder_replica parameter is set in request.
+        // 2.the replica on the to-be-demoted broker is leader replica, the partition is currently under replicated and
+        // skip_under_replicated_partition parameter is set in request.
+        if (!(_skipUnderReplicatedPartition && r.isLeader() && isPartitionUnderReplicated(r))
```

Comment:
What if (1) the replica on a demoted broker is currently a follower and (2) `_skipUnderReplicatedPartition` is `true`? IIUC, there will still be a replica swap, in which this replica moves to the end of replica list in the corresponding follower.

Can we also handle this scenario?

Reply:
Good point. I agree `_skipUrpDemotion`(original `_skipUnderReplicatedPartition`) should cover this scenario.